lib/utils/aws/credentials.go

```diff
@@ -0,0 +1,147 @@
+	return &credentialsGettter{}
+}
+
+// Get obtains STS credentials.
+func (g *credentialsGettter) Get(_ context.Context, request GetCredentialsRequest) (*credentials.Credentials, error) {
+	logrus.Debugf("Creating STS session %q for %q.", request.SessionName, request.RoleARN)
+	return stscreds.NewCredentials(request.Provider, request.RoleARN,
+		func(cred *stscreds.AssumeRoleProvider) {
+			cred.RoleSessionName = request.SessionName
+			cred.Expiry.SetExpiration(request.Expiry, 0)
+
+			if request.ExternalID != "" {
+				cred.ExternalID = aws.String(request.ExternalID)
+			}
+		},
+	), nil
+}
+
+// CachedCredentialsGetterConfig is the config for creating a CredentialsGetter that caches credentials.
+type CachedCredentialsGetterConfig struct {
+	// Getter is the CredentialsGetter for obtaining the STS credentials.
+	Getter CredentialsGetter
+	// CacheTTL is the cache TTL.
+	CacheTTL time.Duration
+	// Clock is used to control time.
+	Clock clockwork.Clock
+}
+
+// SetDefaults sets default values for CachedCredentialsGetterConfig.
+func (c *CachedCredentialsGetterConfig) SetDefaults() {
+	if c.Getter == nil {
+		c.Getter = NewCredentialsGetter()
+	}
+	if c.CacheTTL <= 0 {
+		c.CacheTTL = time.Minute
+	}
+	if c.Clock == nil {
+		c.Clock = clockwork.NewRealClock()
+	}
+}
+
+type cachedCredentialsGetter struct {
+	config CachedCredentialsGetterConfig
+	cache  *utils.FnCache
+}
+
+// NewCachedCredentialsGetter returns a CredentialsGetter that caches credentials.
+func NewCachedCredentialsGetter(config CachedCredentialsGetterConfig) (CredentialsGetter, error) {
+	config.SetDefaults()
+
+	cache, err := utils.NewFnCache(utils.FnCacheConfig{
+		TTL:   config.CacheTTL,
+		Clock: config.Clock,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &cachedCredentialsGetter{
+		config: config,
+		cache:  cache,
+	}, nil
+}
+
+// Get returns cached credentials if found, or fetch it from the configured
+// getter.
+func (g *cachedCredentialsGetter) Get(ctx context.Context, request GetCredentialsRequest) (*credentials.Credentials, error) {
+	credentials, err := utils.FnCacheGet(ctx, g.cache, request, func(ctx context.Context) (*credentials.Credentials, error) {
+		credentials, err := g.config.Getter.Get(ctx, request)
+		return credentials, trace.Wrap(err)
+	})
+	return credentials, trace.Wrap(err)
+}
+
+type staticCredentialsGetter struct {
+	credentials *credentials.Credentials
+}
+
+// NewStaticCredentialsGetter returns a CredentialsGetter that always returns
+// the same provided credentials.
+//
+// Used in testing to mock CredentialsGetter.
+func NewStaticCredentialsGetter(credentials *credentials.Credentials) CredentialsGetter {
+	return &staticCredentialsGetter{
+		credentials: credentials,
+	}
+}
+
+// Get returns the credentials provided to NewStaticCredentialsGetter.
+func (g *staticCredentialsGetter) Get(_ context.Context, _ GetCredentialsRequest) (*credentials.Credentials, error) {
+	return g.credentials, nil
+}
```

Comment:
It is possible to pass nil as credentials. WDYT about checking for nil credentials in `New...()` or, alternatively, returning error in `Get()`?